plugins/calendar-action/src/index.ts

```diff
@@ -0,0 +1,95 @@
+//
+// Copyright © 2020 Anticrm Platform Contributors.
+//
+// Licensed under the Eclipse Public License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License. You may
+// obtain a copy of the License at https://www.eclipse.org/legal/epl-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import type { Ref, TxUpdateDoc } from '@anticrm/core'
+import core, { DeferredPromise } from '@anticrm/core'
+import type { CalendarService, Event } from '@anticrm/calendar'
+import { setResource } from '@anticrm/platform'
+import calendar from '@anticrm/calendar'
+import { Action } from '@anticrm/action'
+
+const wait = async <T>(p: Promise<T>, delay: number): Promise<null | T> => {
+  if (delay <= 0) {
+    return null
+  }
+
+  let tid: number | undefined
+  let delayResolve = (x: null): void => {}
+  const res = await Promise.race([
+    p,
+    new Promise<null>(resolve => {
+      delayResolve = resolve
+      tid = setTimeout(() => resolve(null), delay) as unknown as number
+    })
+  ])
+
+  if (res !== null) {
+    delayResolve(null)
+
+    if (tid !== undefined) {
+      clearTimeout(tid)
+    }
+  }
+
+  return res
+}
+
+const waitForEvent = new Action()
+  .exec(
+    async (ctx) => {
+      const eventID: Ref<Event> = ctx.pop()
+      const event = (await ctx.findAll(calendar.class.Event, { _id: eventID }))[0]
+
+      if (event === undefined) {
+        throw Error(`Event is not found: ${eventID}`)
+      }
+
+      let { endsAt } = event
+      let update = new DeferredPromise<number>()
+
+      const unsub = ctx.subscribe({ id: event._id }, async (tx) => {
+        if (tx._class !== core.class.TxUpdateDoc) {
```

Comment:
What if event is removed? It will wait forever?

Reply:
Planned to handle this case in different PR. In general this case is complex, it is necessary to allow user to cancel/create new instance of action, and removing or archiving event would lead to action cancel or something like that.